write flight datasets to zarr files

yoctoyotta1024 opened the pull request:
- hardcodes chunk sizes of datasets to be more optimal considering the typical lengths of the dimensions of `cocx, range, fft` and `iq`. Fixed chunk sizes also makes it possible to write the dataset(s) to .zarr files (`chunks="auto"` makes the chunk sizes irregular)
- add option to write datasets from a flight using the flightname given the directory where they should be written (do we want this?)

yoctoyotta1024 committed:
feat: add option to write flight datasets to zarr files

yoctoyotta1024 committed:
refactor: hardcode dask chunk sizes

to be more optimal given dimension lengths and to be compatible with writing zarr datasets

d70-t commented on hamp_radar/flight_dataset.py:
```diff
@@ -36,7 +36,13 @@ def read_datasetblock(
-    dataset = [read_datasetblock(d, ds_ppar).chunk("auto") for d in dsgeom.data]
+    chunks = {
+        "cocx": 2,
+        "range": 512,  # TODO(ALL): see HACK about range dimension = 512
+        "fft": 256,
```
These are likely good chunks for reading the pds, but we might want to bunch up a few timesteps when writing. Probably we also want to split at least co and cross channels (often only co is used) and we may want to split range.

We also need to ensure, that coordinate variables (especially 1D) are preferentially stored as a single chunk (e.g. time).

yoctoyotta1024 replied:
- for splitting the co and cross channels and splitting range we can do that when writing really easily (I can add that)
- for bunching time I suggest we set the chunks of frame when reading to some value so that the overall chunk stays around 100MB, so maybe chunks `{"frame": 64, ...}` . Is that also what you had in mind?
- I'm not sure how to make coordinate variables be stored as a single chunk but I'll look into it

d70-t replied:
Essentially, for writing, you'd want to define the chunking (and other things) in the `encoding` argument which you can pass to `to_zarr`. It's a dictionary containing individual settings for each variable.
The chunks here are `dask` chunks, which are in-memory chunks. It's best if the dask chunks align both, with the source data chunk pattern and the target chunk data pattern.

yoctoyotta1024 replied:
I'm sorry I can't see how to put this into practise. I think I'm mistaken somewhere and it'd be better to discuss this in person, but here is what I understand so far:

- We then want all 1-D variables to be written as one chunk, which means they must already be one chunk when the dataset is created using dask. The 1-D variables had dim=frame so we need the frame dimension in the chunks dict when we concatenate the datasets to be at least the length of the concatenated dataset's frame dimension?? But this would have to be known before creating the concatenated dataset?? I don't think I understand this correctly.
- We also want to make coordinate variables stored as one chunk, but iq is a coordinate and is 4-D and massive, so making it one chunk seems like a very bad idea to me?
- I thought we want the zarr chunks to be around 10MB and so I thought we should make the frame chunk size the appropriate size such that when we make zarr chunks (with the co and cross channels and the range dimensions split) they would end up being about 10MB?

yoctoyotta1024 replied:
Okay so based on what we discussed on Saturday I came up with a draft for how to chunk and re-chunk the dask arrays and then write them to smaller chunks in zarr. Let me know what you think! :)

d70-t reviewed (approved):
This sounds great.👍

yoctoyotta1024 committed:
feat: new file for writing dataset functions

yoctoyotta1024 commented on hamp_radar/writezarr.py:
```diff
@@ -0,0 +1,67 @@
+def write_iqdataset(ds: xr.Dataset, filename: Path):
+    encoding = zarr_encoding(ds)
+    ds = prepare_daskarrays_for_zarrchunks(ds, encoding)
+    assert good_chunksizes(ds, maxchunksize_mbytes=10000), "some dask chunks are > 10GB"
```
I think this is overkill and we can delete this to be honest

yoctoyotta1024 committed:
refactor: allow option to merge chunks encoding to an existing encoding dict

d70-t commented on hamp_radar/writezarr.py:
```diff
@@ -0,0 +1,83 @@
+    return encoding
+
+
+def merge_encodings(e1: dict, e2: dict):
+    for x in set(e1.keys()) & set(e2.keys()):
+        e1[x].update(e2[x])
+    for x in set(e2.keys()) - set(e1.keys()):
+        e1[x] = e2[x]
+    return e1
```
This function modifies `e1`, which is potentially dangerous and should be avoided.

yoctoyotta1024 replied:
Better to create a new dictionary? Or we could raise an error instead if any existing encodings exist instead of calling `merge_encodings`?

d70-t replied:
Having a function to merge encodings might be fine (although at the moment I don't see all the implications it might have in my head). But anyways, this function shouldn't modify its inputs but rather return a new dictionary. I'd probably go with a construction similar to:

```python
return {**e1, **e2}
```

This might however need some adaptation, depending on how nested keys should be merged.

yoctoyotta1024 replied:
I think this way deals with the merging appropriately: https://github.com/halo-narval/hamp_radar/pull/18/commits/73c68742f7f839ab2ea785a349d55da01f82bd3b
it creates a new dictionary with keys for each variable from both e1 and e2, but it would throw an error if e1 and e2 had matching keys, e.g. if you tried to set chunks for a variable in both e1 and e2.

d70-t commented on hamp_radar/writezarr.py:
```diff
@@ -0,0 +1,83 @@
+
+
+def prepare_daskarrays_for_zarrchunks(ds: xr.Dataset, encoding: dict):
+    def safe_chunks(current_chunks, desired_chunks):
+        return [max(i, c) for i, c in zip(current_chunks, desired_chunks)]
```
What exactly is the purpose of `safe_chunks`? It it suppost to find a good common chunking for the combination of `current_chunks` and `desired_chunks`? If so, you may want to search for the [least common multiple](https://en.wikipedia.org/wiki/Least_common_multiple).

yoctoyotta1024 replied:
Oh I was aiming to make the new chunks at least as big as the `desired_chunks`, but now I see that's wrong and actually we want the least common multiple (I didn't consider the case when it wasn't assumed that `desired_chunks` was already a factor/multiple of the `current_chunks`). I'll redo this.

d70-t commented on hamp_radar/writezarr.py:
```diff
@@ -0,0 +1,83 @@
+
+def write_iqdataset(ds: xr.Dataset, filename: Path):
+    encoding = zarr_encoding(ds)
+    ds = prepare_daskarrays_for_zarrchunks(ds, encoding)
```
At this point, it's probably a bit too late to find a "good" common chunking. If you want to read the data at a reasonable common chunking, you want to do this in the moment where you convert from memmapped numpy arrays to dask arrays. Once you have dask arrays, it's a bit too late to "remove" the initial chunking, and you might not gain that much anymore from optimizing it.

yoctoyotta1024 replied:
Ah okay, I'll change this then and make the chunking when we write the datasets the desired chunking rather than the dictionary it is currently.

d70-t commented on hamp_radar/flight_dataset.py:
```diff
@@ -36,7 +37,14 @@ def read_datasetblock(
+        "iq": 2,
+        "frame": 128,  # TODO(ALL): HACK makes iq dask array chunks circa. 100MB
+    }
+    dataset = [read_datasetblock(d, ds_ppar).chunk(chunks) for d in dsgeom.data]
```
This is where you define the initial dask chunking. The chunking at this point should be at least close to the desired (`safe_chunks`?) chunksize.

yoctoyotta1024 replied:
Yes exactly what I thought from your last comment too! Will redo.

yoctoyotta1024 replied:
When combining the datasets from each pds file into one final dataset the frame dimension needs to be rechunked (again). I therefore restructed these functions so they better subdivide the tasks.

It looks more complicated than it is because I left in the ability to use dask's "auto" chunks instead of our `ideal_daskchunks` (via the `use_autochunks` boolean) but if you agree I would remove that option.

yoctoyotta1024 committed:
refactor: option to set chunks for each variable individually based on their dimensions when creating a dataset

so that each array's chunks are more ideally shaped (e.g. for future writing in zarr format) and each chunk circa. 250MB.
alternatively, dasks auto chunking is used

refactor: option to rechunk frame dimension when concatenating

refactor: reword functions and remove dask rechunking before writing

yoctoyotta1024 committed:
refactor: main function shows option of using autochunks when not writing dataset

d70-t commented on hamp_radar/writezarr.py:
```diff
@@ -0,0 +1,84 @@
+        "cocx": 2,
+    },
+    ("frame", "range", "cocx"): {
+        "frame": 512000,
```
This number is not a multiple of 81920 (i.e. the chunk size of `ideal_zarrchunks`). This could probably create problems when writing datasets.

yoctoyotta1024 replied:
Ah but I thought since it is a multiple of 32 it is okay (since 81920 is the outermost dimension)? Nevertheless I can change it to be safe, or perhaps I've misunderstood the chunking

yoctoyotta1024 replied:
Ah okay, I think I misunderstood how chunking works since I thought all the dimensions had to be multiples of each other except for the outermost one. I've corrected it now to ```"frame": 419520,```  (almost 512000) so that dask chunks are exactly 6x the size of zarr ones along the "frame" dimension here.

d70-t commented:
> @d70-t can we merge this? It might be useful to have it in the main branch already.

I guess that's fine, sorry for not getting back earlier. You might want to check [this comment first](https://github.com/halo-narval/hamp_radar/pull/18#discussion_r1958221876).

yoctoyotta1024 commented:
> > @d70-t can we merge this? It might be useful to have it in the main branch already.
> 
> I guess that's fine, sorry for not getting back earlier. You might want to check [this comment first](https://github.com/halo-narval/hamp_radar/pull/18#discussion_r1958221876).

Just replied :) No problem about the delay - I just got reminded of it in a meeting that's why I'm revisiting!

yoctoyotta1024 committed:
fix: frame dimension for ideal dask chunks is multiple of ideal dask chunks

d70-t commented:
Ok, so let's go 👍

yoctoyotta1024 commented:
> Ok, so let's go 👍

Yes please! :D
